crates/typst/src/util/hash.rs

```diff
@@ -0,0 +1,154 @@
+use std::any::Any;
+use std::fmt::{self, Debug};
+use std::hash::{Hash, Hasher};
+use std::ops::{Deref, DerefMut};
+
+use portable_atomic::{AtomicU128, Ordering};
+use siphasher::sip128::{Hasher128, SipHasher13};
+
+/// A wrapper type with lazily-computed hash.
+/// This is useful if you want to pass large values of `T` to memoized
+/// functions. Especially recursive structures like trees benefit from
+/// intermediate prehashed nodes.
+///
+/// Note that for a value `v` of type `T`, `hash(v)` is not necessarily equal to
+/// `hash(LazyHash::new(v))`. Writing the precomputed hash into a hasher's
+/// state produces different output than writing the value's parts directly.
+/// However, that seldomly matters as you are typically either dealing with
+/// values of type `T` or with values of type `LazyHash<T>`, not a mix of both.
+///
+/// # Equality
+/// Because typst uses high-quality 128 bit hashes in all places, the risk of a
+/// hash collision is reduced to an absolute minimum. Therefore, this type
+/// additionally provides `PartialEq` and `Eq` implementations that compare by
+/// hash instead of by value. For this to be correct, your hash implementation
+/// **must feed all information relevant to the `PartialEq` impl to the
+/// hasher.**
+///
+/// # Usage
+/// It is best used inside of an `Arc` or `Rc` to best re-use the hash once it
+/// has been computed.
+pub struct LazyHash<T: ?Sized> {
+    /// The hash for the value.
+    hash: AtomicU128,
+    /// The underlying value.
+    value: T,
+}
+
+impl<T> LazyHash<T> {
+    /// Wrap an item without pre-computed hash.
+    #[inline]
+    pub fn new(value: T) -> Self {
+        Self { hash: AtomicU128::new(0), value }
+    }
+
+    /// Wrap an item with a pre-computed hash.
+    ///
+    /// # Safety
+    /// The hash must be correct for the value. This cannot be enforced at
+    /// compile time, so use with caution.
+    #[inline]
+    pub unsafe fn with_hash(value: T, hash: u128) -> Self {
+        Self { hash: AtomicU128::new(hash), value }
```

Comment:
```suggestion
    /// **Important:** The hash must be correct for the value. This cannot be
    /// enforced at compile time, so use with caution.
    #[inline]
    pub fn with_hash(value: T, hash: u128) -> Self {
        Self { hash: AtomicU128::new(hash), value }
```
I don't think this should be `unsafe`. Unsafe should be reserved for things that can actually cause memory unsafety or undefined behaviour, not logic bugs.